File: src/config/agents-md.ts
import { readdirSync, readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join, resolve } from "node:path";
import { getHeddleHome } from "./paths.ts";

/**
 * Find all AGENTS.md files walking up from startDir to home directory.
 * Case-insensitive matching. Also checks HEDDLE_HOME/AGENTS.md.
 * Returns paths ordered farthest-first → closest-last.
 */
export function findAllAgentsMd(startDir?: string): string[] {
	const start = resolve(startDir ?? process.cwd());
	const home = homedir();
	const found: string[] = [];

	// Walk up from startDir toward home (inclusive), or to filesystem root
	let current = start;
	while (true) {
		const match = findAgentsMdIn(current);
		if (match) {
			found.push(match);
		}

		// Stop at home directory — don't go above it
		if (current === home) break;

		const parent = resolve(current, "..");
		// Stop if we've reached the filesystem root (parent === current)
		if (parent === current) break;

		current = parent;
	}

	// Reverse so farthest-first → closest-last
	found.reverse();

	// Check HEDDLE_HOME
	const heddleHome = getHeddleHome();
	const heddleMatch = findAgentsMdIn(heddleHome);
	if (heddleMatch && !found.includes(heddleMatch)) {
		// Prepend HEDDLE_HOME (it's the most global scope)
		found.unshift(heddleMatch);
	}

	return found;
}

/**
 * Look for an AGENTS.md file (case-insensitive) in a directory.
 * Returns the full path if found, undefined otherwise.
 */
function findAgentsMdIn(dir: string): string | undefined {
	try {
		const entries = readdirSync(dir);
		const match = entries.find((entry) => entry.toLowerCase() === "agents.md");
		return match ? join(dir, match) : undefined;
	} catch {
		return undefined;
	}
}

/**
 * Load and concatenate all AGENTS.md files.
 * Returns null if no files found.
 */
export function loadAgentsContext(startDir?: string): string | null {
	const paths = findAllAgentsMd(startDir);
	if (paths.length === 0) return null;

	return paths.map((p) => readFileSync(p, "utf-8")).join("\n\n");
}

File: src/config/agents_md.rs
//! AGENTS.md discovery and concatenation.

use std::path::{Path, PathBuf};

use super::paths::get_heddle_home;

fn find_agents_md_in(dir: &Path) -> Option<PathBuf> {
    let entries = std::fs::read_dir(dir).ok()?;
    for entry in entries.flatten() {
        if let Some(name) = entry.file_name().to_str() {
            if name.to_lowercase() == "agents.md" {
                return Some(dir.join(name));
            }
        }
    }
    None
}

pub fn find_all_agents_md(start_dir: Option<&Path>) -> Vec<PathBuf> {
    let cwd_storage;
    let start = match start_dir {
        Some(d) => d.to_path_buf(),
        None => {
            cwd_storage = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
            cwd_storage.clone()
        }
    };
    let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/"));

    let mut found = Vec::new();
    let mut current = start;
    loop {
        if let Some(p) = find_agents_md_in(&current) {
            found.push(p);
        }
        if current == home {
            break;
        }
        match current.parent() {
            Some(parent) if parent != current => current = parent.to_path_buf(),
            _ => break,
        }
    }

    found.reverse();

    let heddle_home = get_heddle_home();
    if let Some(p) = find_agents_md_in(&heddle_home) {
        if !found.contains(&p) {
            found.insert(0, p);
        }
    }

    found
}

pub fn load_agents_context(start_dir: Option<&Path>) -> Option<String> {
    let paths = find_all_agents_md(start_dir);
    if paths.is_empty() {
        return None;
    }
    let parts: Vec<String> = paths
        .iter()
        .filter_map(|p| std::fs::read_to_string(p).ok())
        .collect();
    if parts.is_empty() {
        None
    } else {
        Some(parts.join("\n\n"))
    }
}
